src/nodes/flow/coupling_flows/planar_flow.jl

```diff
@@ -0,0 +1,258 @@
+
+# jacobian of the PlanarFlow function (univariate input)
+function _jacobian(f::PlanarFlow{T1,T2}, input::T3) where { T1 <: Real, T2 <: Real, T3 <: Real } 
+
+    # fetch values 
+    u, w, b = getall(f)
+
+    # calculate result (optimized)
+    result = u * w * dtanh(w * input + b) + 1
+
+    # return result
+    return result
+
+end
+jacobian(f::PlanarFlow{T1,T2}, input::T3) where { T1 <: Real, T2 <: Real, T3 <: Real } = _jacobian(f, input)
+Broadcast.broadcasted(::typeof(jacobian), f::PlanarFlow{T1,T2}, input::Array{T3,1}) where { T1 <: Real, T2 <: Real, T3 <: Real } = broadcast(_jacobian, Ref(f), input)
+
+function _jacobian(f::PlanarFlow{T1,T2}, input::T3) where { T1 <: Real, T2 <: Real, T3 }
+    # function when the input is an array with 1 element
+    @assert length(input) == 1 "Something is wrong with the dimensionality of the input to the PlanarFlow flow."
+    return jacobian(f, input[1])
+end
+jacobian(f::PlanarFlow{T1,T2}, input::T3) where { T1 <: Real, T2 <: Real, T3 } = _jacobian(f, input)
+Broadcast.broadcasted(::typeof(jacobian), f::PlanarFlow{T1,T2}, input::Array{T3,1}) where { T1 <: Real, T2 <: Real, T3 } = broadcast(_jacobian, Ref(f), input)
+
+# inplace jacobian of the PlanarFlow function (multivariate input)
+function jacobian!(output::Array{T2,2}, f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real}
+
+    # check whether the dimensionality is correct
+    @assert size(output) == (length(input), length(f.u)) "The dimensionality of the preallocated jacobian matrix seems incorrect."
+
+    # fetch values 
+    u, w, b = getall(f)
+
+    # calculate result
+    for ku = 1:length(u)
+        for kw = 1:length(w)
+            output[ku,kw] = u[ku]*w[kw]
+        end
+    end
+    output .*= dtanh(dot(w, input) + b)
+    @inbounds for k = 1:length(input)
+        output[k,k] += 1.0
+    end
+
+end
+
+# determinant of the jacobian of the PlanarFlow function (multivariate input)
+det_jacobian(f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real} = det(jacobian(f, input))
+
+# determinant of the jacobian of the PlanarFlow function (univariate input)
+function det_jacobian(f::PlanarFlow{T, T}, input::T) where { T <: Real }
+
+    # fetch values
+    u, w, b = getall(f)
+
+    # calculate result
+    result = 1 + dot(u, w)*dtanh(dot(w, input) + b)
+
+    # return result
+    return result
+
+end
+
+# extra utility function (multivariate)
+inv_jacobian(f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real }       = inv(jacobian(f, input))
+absdet_jacobian(f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real }    = abs(det_jacobian(f, input))
+logabsdet_jacobian(f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real}  = logabsdet(jacobian(f, input))
+
+# extra utility functions (univariate)
+inv_jacobian(f::PlanarFlow{T,T}, input::T) where { T <: Real }               = 1.0 / jacobian(f, input)
```

Comment:
Be careful here. So I noticed that in many places you use for example `{T1, T2} where { T1 <: Real, T2 <: Real }`. This form accepts `{Float64, Float32}` for example, because both type arguments have different names. However when you type `{T, T} where { T <: Real }` it assumes that both of the **are the same**. So some of your functions accepts different types from `Real` subtype.

For example I do not see any conceptual differences in these
```
inv_jacobian(f::PlanarFlow{T1,T2}, input::T1) where { T1, T2 <: Real }       = inv(jacobian(f, input))

inv_jacobian(f::PlanarFlow{T,T}, input::T) where { T <: Real }               = 1.0 / jacobian(f, input)
```

First version is a more generic version of a second one, but for some reason they have different implementations. I think there should not be two different implementations for `{Float32, Float64}` and `{Float64, Float64}`. Maybe it is a bug and there should be something else? I would just delete versions with `{T, T}` as they are doing exactly the same but just more restrict on their input arguments.

Reply:
Ah, I see now, that `T` is not constrained to be `Real`. They are different indeed, my bad.